### PR TITLE
Arcs personal cloud

### PR DESCRIPTION
**Pull request by cromwellian**
Changes since last demo

Work in progress.

1. use new helm deployed version
2. support serving via a TLS backed domain (e.g. skarabrae) Ingress that maps /clouds/<fingerprint> to NodeVM deployment backends
3. Change use deployments instead of pods, allows easier upgrading docker images 
4. Support auto-scaling clusters. To do this, a free 'node' in the cluster is found,  and the GCP disk is attached to it, then the Kubernetes deployment of the NodeVM is told to bind to that specific node.

TODO:  dismount()/unlock()/lock()  still only partially working
TODO: 'find free node' function just picks at random. In testing, this is insufficient, because you
can still exceed CPU resource scheduling limits on GKE. Instead, this function needs to find a node that can accept another Pod (needs atleast 10% of CPU free)

### CODE REVIEW

File: runtime/ts/keymgmt/webcrypto.ts [hunk 1]
```diff
@@ -191,6 +191,15 @@ class WebCryptoSessionKey implements SessionKey, TestableKey {
 
     }
 
+    export():PromiseLike<string> {
+      return crypto.subtle.exportKey("raw", this.sessionKey).then((raw) => {
+        const buf = new Uint8Array(raw);
+        let res = "";
+        buf.forEach((x) => res += (x < 16 ? '0' : '') + x.toString(16));
```
Comment (shaper) [inline runtime/ts/keymgmt/webcrypto.ts]: Add a comment on what this is doing.

File: server/src/arcs-master-app.ts [hunk 2]
```diff
@@ -32,41 +40,128 @@ class ArcsMasterApp extends AppBase {
...
+    console.log("gcp safe fingerprint " + fingerprint);
+    try {
+      const cloud = CloudManager.forGCP();
+      const disk = await cloud.disks().find(fingerprint);
+      if (disk) {
+        const attached = await disk.isAttached();
+
+        if (attached) {
+          const container: Container | null = await cloud.containers().find(fingerprint);
+          if (container != null) {
```
Comment (shaper) [inline server/src/arcs-master-app.ts]: Perhaps we should log something here (and for unlock case below) if the container is null? Since the disk will not be dismounted, but seems not to in fact be attached to the noted container (or, clarify if I mis-speak).

I see you have a TODO at line 89 below for the unlock case, so I'm wondering what we should do in this case, and probably we should log something in unlock as well.

File: server/src/arcs-master-app.ts [hunk 3]
```diff
@@ -32,41 +40,128 @@ class ArcsMasterApp extends AppBase {
...
     const fingerprint = this.gcpSafeIdentifier(req.params.fingerprint);
-    const wrappedkey = req.params.wrappedkey;
+    const wrappedKey = req.params.wrappedKey;
+    const rewrappedKey = req.params.rewrappedKey;
+
     try {
       const cloud = CloudManager.forGCP();
-      const disk = await cloud.disks().create(fingerprint, wrappedkey);
+      const disk = await cloud.disks().create(fingerprint, wrappedKey, rewrappedKey);
       console.log("Disk successfully created with id " + disk.id());
```
Comment (shaper) [inline server/src/arcs-master-app.ts]: Probably want to null-check (or however errors are reported from `create` and `deploy`) here and line 143 to make sure disk and container were created. Or, do they throw an exception which would hit line 146?

Similar for `mount` implementation below.

File: server/src/arcs-master-app.ts [hunk 4]
```diff
@@ -32,41 +40,128 @@ class ArcsMasterApp extends AppBase {
...
+      const disk = await cloud.disks().find(fingerprint);
+      if (disk) {
+        const attached = await disk.isAttached();
+        const wrappedKey  = await disk.wrappedKeyFor(fingerprint);
+
+        if (attached) {
+          const container: Container | null = await cloud.containers().find(fingerprint);
+          if (container != null) {
+            const status = container.status();
+            if (status !== 'Running') {
```
Comment (shaper) [inline server/src/arcs-master-app.ts]: Here and below is a bit hard to follow due to deep nesting and combination of `else` and `return`, could maybe restructure a bit with helper methods or using either `else` or `return` rather than both, could do later, just mentioning while reading.

File: server/tsconfig.json [hunk 5]
```diff
@@ -3,8 +3,8 @@
   "compilerOptions": {
     "outDir": "./dist",
     "rootDir": "./src",
-    "strictFunctionTypes": true,
-    "strictNullChecks": true
+    "strictFunctionTypes": false,
```
Comment (shaper) [inline server/tsconfig.json]: Why this change here and below? @lindner should confirm.

File: server/src/pouch-db-app.ts [hunk 6]
```diff
@@ -54,10 +56,46 @@ class PouchDbApp extends AppBase {
...
+        "methods": "GET, PUT, POST, HEAD, DELETE"
+      }
+    };
+
+
     if (process.env[ON_DISK_DB]) {
       const dboptions = {prefix: '/personalcloud/'} as PouchDB.Configuration.RemoteDatabaseConfiguration;
       const onDiskPouchDb = PouchDB.defaults(dboptions);
-      this.express.use('/', PouchDbServer(onDiskPouchDb, {mode: 'fullCouchDB', inMemoryConfig: true}));
+      const pouchDbRouter = PouchDbServer(onDiskPouchDb, config);
```
Comment (shaper) [inline server/src/pouch-db-app.ts]: Below seems ok to land PR but further cleanup appears warranted, perhaps file an issue to restructure in whatever manner is appropriate? It looks like there's both potentially more complex LB logic we might want eventually, and some kind of Fauxton issue. Possible both of these are fine as implemented and no follow up issue needed but figured to ask while reading.

File: server/src/deployment/utils.ts [hunk 7]
```diff
@@ -17,4 +25,53 @@ export function arcsKeyFor(fingerprint: string):string {
...
+      } catch (e) {
+        if (!(e instanceof Timeout)) {
+          return Promise.reject(e);
+        }
+      }
+      if (waiting) {
+        try {
+          await delay(timeout);
+        } catch (e) {
+        }
```
Comment (shaper) [inline server/src/deployment/utils.ts]: Probably want to log something if we hit an exception?

File: server/src/pouch-db-app.ts [hunk 8]
```diff
@@ -54,10 +56,46 @@ class PouchDbApp extends AppBase {
...
   private addPouchRoutes(): void {
+    // If VM lives at non-root prefix, works but fauxton not fully working yet
+    const urlPrefix = process.env[VM_URL_PREFIX] || '/';
+    const config = {
+      mode: 'fullCouchDB', inMemoryConfig: true,
+      "httpd": {
+        "enable_cors": true
+      },
+      "cors": {
+        "origins": "https://skarabrae.org",
```
Comment (shaper) [inline server/src/pouch-db-app.ts]: General comment, the server code has a mix of config files and config in TS now (here's one example but there's also lots of GCP specific bits in TS constants). Likely worth moving everything to config YAML or otherwise at some point, I saw you had similar comment about doing things in JSON instead elsewhere. Perhaps worth a follow up issue filing.

File: Dockerfile [hunk 9]
```diff
@@ -20,6 +20,7 @@ WORKDIR /usr/src/app/server
 RUN npm install --no-package-lock && npm install --package-lock-only
 
 RUN npm test
+# RUN npm run build
```
Comment (lindner) [inline Dockerfile]: revert?

File: package.json [hunk 10]
```diff
@@ -22,11 +22,11 @@
     "build:typedoc": "typedoc --mode file --target ES5  --downLevelIteration -out dist/apidocs"
   },
   "devDependencies": {
-    "@types/chai": "^4.1.6",
+    "@types/chai": "^4.1.7",
```
Comment (lindner) [inline package.json]: probably need to rebase this"

File: server/tsconfig.json (hunk 5 above)
Comment (cromwellian) [reply server/tsconfig.json]: Bad merge, this comes from before Paul flipped 'em, will revert.

**Comment by cromwellian** [comment]
> Feels like it's approaching the point where it's worth adding some unit tests.

Agreed. The only reason I haven't attempted it yet is I need to come up with a good harness to test this stuff. Spinning up a GCP cluster for hermetic tests is too slow and expensive. 

The most likely solution will be to use Minikube (https://kubernetes.io/docs/setup/minikube/) to run a local version of the container engine environment, however we have a dependency on the GCP encrypted disk infrastructure, so I'll need to mock that out, or implement a Cryptfs version, the latter seems more likely, since it would also be useful as a scalable alternative to GCP disks (e.g. CryptFS on top of Bucket Storage)

I'll do this in a follow-on after next weeks demos.

**Comment by shaper** [comment]
Checking on this PR, is it ready to merge once above comments are addressed? Would be good to land this (and the PWA).

**Comment by cromwellian** [comment]
I will merge it soon, just doing some last minute end-to-end verifications of the server, provisioning, and android.

File: server/src/arcs-master-app.ts (hunk 3 above)
Comment (cromwellian) [reply server/src/arcs-master-app.ts]: Currently, it's exception based. I added an intermediate check and a TODO to add better error handling.

File: server/src/arcs-master-app.ts (hunk 4 above)
Comment (cromwellian) [reply server/src/arcs-master-app.ts]: I restructed with early-exits

File: server/src/deployment/utils.ts (hunk 7 above)
Comment (cromwellian) [reply server/src/deployment/utils.ts]: I realize it's confusing, but the Promise rejection is expected. The delay() function returns a Promise that is never resolved, only rejected after it's timeout expires. 

This is used in Promise.race(promise, delay) to detect a timeout where the first promises take too long. It is reused in this fragment to actually wait. I could make a version of delay() that resolves after timeout to avoid the try/catch trick, or hide this try/catch block inside of a helper which I'll do for now.

File: server/src/pouch-db-app.ts (hunk 8 above)
Comment (cromwellian) [reply server/src/pouch-db-app.ts]: Created issue

File: package.json (hunk 10 above)
Comment (cromwellian) [reply package.json]: Done
